pkg/controller/remotemachineset/awsactuator.go

```diff
@@ -348,34 +351,62 @@ func (a *AWSActuator) getSubnetsByAvailabilityZone(pool *hivev1.MachinePool) (ma
 		return nil, err
 	}
 
-	conflictingSubnets := sets.NewString()
-	subnetsByAvailabilityZone := make(map[string]string, len(pool.Spec.Platform.AWS.Subnets))
+	vpc := *results.Subnets[0].VpcId
+	if vpc == "" {
+		return nil, errors.Errorf("%s has no VPC", *results.Subnets[0].SubnetId)
+	}
+
+	routeTables, err := a.awsClient.DescribeRouteTables(&ec2.DescribeRouteTablesInput{
+		Filters: []*ec2.Filter{{
+			Name:   aws.String("vpc-id"),
+			Values: []*string{aws.String(vpc)},
+		}},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "describing route tables")
+	}
+
+	var privateSubnets, publicSubnets = map[string]ec2.Subnet{}, map[string]ec2.Subnet{}
 	for _, subnet := range results.Subnets {
-		if subnetID, ok := subnetsByAvailabilityZone[*subnet.AvailabilityZone]; ok {
-			conflictingSubnets.Insert(*subnet.SubnetId)
-			conflictingSubnets.Insert(subnetID)
-			continue
+		isPublic, err := isSubnetPublic(routeTables.RouteTables, *subnet.SubnetId, a.logger)
+		if err != nil {
+			return nil, errors.Wrap(err, "describing route tables")
+		}
+		if isPublic {
+			publicSubnets[*subnet.SubnetId] = *subnet
+		} else {
+			privateSubnets[*subnet.SubnetId] = *subnet
 		}
-		subnetsByAvailabilityZone[*subnet.AvailabilityZone] = *subnet.SubnetId
 	}
 
-	if len(conflictingSubnets) > 0 {
+	if len(publicSubnets) > 0 {
```

Comment:
What happens if we don't have this condition check?

Reply:
Public subnets are not mandatory to be mentioned, so it proceeds to fetch private subnets (which must be mentioned if using subnets) if public subnets are not present

Reply:
should we have same check for private subnets?

Reply:
Private subnets are mandatory (if subnets are mentioned)... And since the total subnets are Public+Private, we don't need a check for that